Name Expression Preview Test Automation

labkey-danield opened the pull request:
#### Rationale
Added test automation to test the preview functionality of the sample name expression (naming pattern).

#### Related Pull Requests
* None.

#### Changes
* Added preview component to page.
* Added test to verify the preview functionality of naming patters for samples.

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Update DomainDesignerPage to have tooltip component for name expression preview.
Add test for name expression preview.

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Made DomainDesignerPage tooltip component be more 'universal'.
Added tests to validate name expression with a source parent (LKSM only).
Added tests to validate name expression with aliquots (LKSM only).

XingY commented on src/org/labkey/test/tests/SampleTypeNameExpressionTest.java:
```diff
@@ -232,4 +286,298 @@ private void verifyNames(String sampleTypeName, String header, String nameExpres
+     * @throws Exception Can be thrown by test helper.
+     */
+    @Test
+    public void testCreateDerivedThroughUI() throws Exception
```
Perhaps rename the test and also modify comments to mention "lineage lookup field". I don't think issue 44760 has to do with creating samples from UI, but rather, it has to do with the latest lineage lookup name expression feature.

labkey-danield replied:
I updated the function name to `testDeriveSampleFromSampleDetailsPage`.
As far as I can tell all the other tests in this file use import to create derived samples, this is the only one (in this test class) that uses the UI. Because of that I wanted to make sure the test name conveys that info.
I have added clarify text to the issue comments.

XingY commented on src/org/labkey/test/tests/SampleTypeNameExpressionTest.java:
```diff
@@ -232,4 +286,298 @@ private void verifyNames(String sampleTypeName, String header, String nameExpres
+        SampleTypeHelper sampleHelper = new SampleTypeHelper(this);
+
+        final String sampleType = "DerivedUI_SampleType";
+        final String nameExpression = String.format("DUI_${genId}_${materialInputs/%s/Str}", PARENT_SAMPLE_TYPE);
```
Once the fix is in, can also add a test to use parent alias here.

labkey-danield replied:
I added a TODO comment and another link to the issue.

XingY commented on src/org/labkey/test/tests/SampleTypeNameExpressionTest.java:
```diff
@@ -232,4 +286,298 @@ private void verifyNames(String sampleTypeName, String header, String nameExpres
+
+            expectedToolTip.append(expectedPreview);
+            expectedToolTip.append("\n");
+        }
```
if there is error, wouldn't there be "Unable to generate example name from the current pattern. Check for syntax errors." msg?

labkey-danield replied:
Yes it does.
In that case the error message is the value sent in. Example:
`String expectedMsg = generateExpectedToolTip("Unable to generate example name from the current pattern. Check for syntax errors.", true);`
The testErrorsAndWarning test has more examples.
I will update the comment to clarify.

XingY replied:
hmm, is there any reason to not have the function check for the error string without having to pass as param? I think this is a static msg that doesn't change based on error reason?

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Moved NameExpression tests out of SampleTypeDesignerTest and into its own test class.
Added SMAliquotCreateNaeExpressionTest.
Added tests from code review feedback.
Made the getPreview component in the designer page to be more generic.

labkey-danield committed:
Merge branch 'develop' into fb_nameExpressionAutomation

labkey-danield committed:
Changed error test, expected error message is now hard coded and not a constructed string.
